File: problem-2_decodeString.java
// Time Complexity :
//      total time complexity will be O(n) where n is the length of the string
//
// Space Complexity :
//      O(n)->where n is the length of the string
//
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

class Solution {
    public String decodeString(String s) {
        if(s.length()==0)
            return "";
        String result="", substring="";
        int multiplier=0, num=0;
        char c;
        Stack<String> stringStack = new Stack<>();
        Stack<Integer> integerStack = new Stack<>();
        for(int i=0;i<s.length();i++)
        {
            c = s.charAt(i);
            if(Character.isDigit(c))
            {
                num = num*10 + c - '0';
            }
            else if(c == '[')
            {
                integerStack.push(num);
                stringStack.push(substring);
                num = 0;
                substring = "";
            }
            else if(c==']')
            {
                multiplier = integerStack.pop();
                StringBuilder str = new StringBuilder();
                for(int j=0; j<multiplier;j++)
                {
                    str.append(substring);
                }
                substring = stringStack.pop();
                substring += str;
            }
            else
            {
                substring =  substring+c;
            }
        }
        return substring;
    }
}

File: Problem2_DecodeString.py
# Time Complexity: O(nk), where n is the length of the string and k - max num in the string
# Space Complexity: O(nk)
# Did this code successfully run on Leetcode: Yes

# Solution:

class Solution:
    def decodeString(self, s: str) -> str:
        if not s or len(s) == 0:
            return ""

        num_stack = []
        str_stack = []
        num = 0
        curr_str = ""

        for ch in s:
            if ch.isdigit():
                num = num * 10 + int(ch)

            elif ch == "[":
                # Push num and curr str to the stacks and reset them
                num_stack.append(num)
                str_stack.append(curr_str)
                num = 0
                curr_str = ""

            elif ch == "]":
                # Repeat curr str n times from num_stack
                n = num_stack.pop()
                curr_str = curr_str * n
                # Append prev str from str_stack
                curr_str = str_stack.pop() + curr_str

            else:
                curr_str = curr_str + ch

        return curr_str

